File: 22-Strategy/src/java/chapter22/CryptoStrategy.java
package chapter22;

/**
 * Concrete Strategy — pays with cryptocurrency.
 */
public class CryptoStrategy implements PaymentStrategy {
    private final String wallet;

    public CryptoStrategy(String wallet) {
        this.wallet = wallet;
    }

    @Override
    public void pay(double amount) {
        System.out.printf("  Paid $%.2f in crypto from wallet %s%n", amount, wallet);
    }
}

File: 22-Strategy/src/rust/crypto_strategy.rs
use crate::payment_strategy::PaymentStrategy;

/// Concrete Strategy — pays with cryptocurrency.
pub struct CryptoStrategy {
    wallet: String,
}

impl CryptoStrategy {
    pub fn new(wallet: &str) -> Self {
        CryptoStrategy { wallet: wallet.to_string() }
    }
}

impl PaymentStrategy for CryptoStrategy {
    fn pay(&self, amount: f64) {
        println!("  Paid ${:.2} in crypto from wallet {}", amount, self.wallet);
    }
}
